File: src/native/androidProject/app/src/main/java/io/yourname/androidproject/camera/HoldController.kt
package io.yourname.androidproject.camera

import android.os.Handler
import android.os.Looper
import android.util.Log

/**
 * Manages the QR hold state — after each detection, results are suppressed for HOLD_DURATION_MS
 * to prevent the same QR from firing repeatedly and to let auto-zoom relax.
 *
 * Uses a suppress flag in BarcodeDetector instead of unbind/rebind — camera pipeline
 * stays bound the whole time, so there is no flicker.
 */
class HoldController(
    private val stateMachine: VideoStreamStateMachine,
    private val barcodeDetector: BarcodeDetector
) {

    private val TAG = "HoldController"
    private val HOLD_DURATION_MS = 200L

    private val handler = Handler(Looper.getMainLooper())
    private var resumeRunnable: Runnable? = null

    var lastDetectedValue: String? = null
        private set

    /**
     * Call after a new QR value is confirmed. Suppresses results for HOLD_DURATION_MS,
     * then re-enables and transitions back to STREAMING.
     */
    fun startHold() {
        if (!stateMachine.transition(VideoStreamState.HOLD)) return

        barcodeDetector.suppressResults = true
        Log.d(TAG, "Hold started — results suppressed for ${HOLD_DURATION_MS}ms")

        val runnable = Runnable {
            if (!stateMachine.state.isActive) return@Runnable
            barcodeDetector.suppressResults = false
            stateMachine.transition(VideoStreamState.STREAMING)
            Log.d(TAG, "Hold ended — results resumed")
        }
        resumeRunnable = runnable
        handler.postDelayed(runnable, HOLD_DURATION_MS)
    }

    /**
     * Call on stop() or flip() to cancel any pending hold and clear detection memory.
     */
    fun reset() {
        resumeRunnable?.let { handler.removeCallbacks(it) }
        resumeRunnable = null
        barcodeDetector.suppressResults = false
        lastDetectedValue = null
        Log.d(TAG, "Hold state reset")
    }

    /**
     * Returns true if this value is new (not a repeat of the last detected value).
     * Also updates lastDetectedValue if it is new.
     */
    fun isNewValue(value: String): Boolean {
        if (value == lastDetectedValue) return false
        lastDetectedValue = value
        return true
    }
}

File: src/native/iosnativeWebView/Sources/Core/Camera/HoldController.swift
import Foundation
import os

private let logger = Logger(subsystem: Bundle.main.bundleIdentifier ?? "com.app", category: "HoldController")

/// Manages QR hold state — after each detection, results are suppressed for HOLD_DURATION
/// to prevent the same QR from firing repeatedly.
/// Uses a suppressResults flag on BarcodeDetector instead of stopping the session,
/// so the camera pipeline stays live and there is no flicker.
class HoldController {

    private let holdDuration: TimeInterval = 0.2  // 200ms, matches Android

    private let stateMachine: VideoStreamStateMachine
    private let barcodeDetector: BarcodeDetector
    private var holdWorkItem: DispatchWorkItem?
    private let serialQueue = DispatchQueue(label: "com.catalyst.camera.hold")

    private(set) var lastDetectedValue: String?

    init(stateMachine: VideoStreamStateMachine, barcodeDetector: BarcodeDetector) {
        self.stateMachine = stateMachine
        self.barcodeDetector = barcodeDetector
    }

    /// Call after a new QR value is confirmed.
    func startHold() {
        guard stateMachine.transition(to: .hold) else { return }

        barcodeDetector.suppressResults = true
        logger.debug("Hold started — results suppressed for \(self.holdDuration * 1000, format: .fixed(precision: 0))ms")

        let workItem = DispatchWorkItem { [weak self] in
            guard let self, self.stateMachine.isActive else { return }
            self.barcodeDetector.suppressResults = false
            self.stateMachine.transition(to: .streaming)
            logger.debug("Hold ended — results resumed")
        }
        holdWorkItem = workItem
        serialQueue.asyncAfter(deadline: .now() + holdDuration, execute: workItem)
    }

    /// Call on stop() or flip() to cancel any pending hold and clear detection memory.
    func reset() {
        serialQueue.async { [weak self] in
            guard let self else { return }
            self.holdWorkItem?.cancel()
            self.holdWorkItem = nil
            self.barcodeDetector.suppressResults = false
            self.lastDetectedValue = nil
            logger.debug("Hold state reset")
        }
    }

    /// Returns true if this value is new (not a repeat of the last detected value).
    func isNewValue(_ value: String) -> Bool {
        serialQueue.sync {
            if value == lastDetectedValue { return false }
            lastDetectedValue = value
            return true
        }
    }
}
